README.md
# Asgard API Plugin Metrics Mesos

Esse projeto implementa a interface de plugin da Asgard API e adiciona novs endpoints que retornam
números sobre o cluster de Mesos.

# Interface de Plugins da Asgard API

A interface de plugins consiste em implementar um método com a seguinte assinatura:

```python
def asgard_api_plugin_init(**kwargs):
  pass
```

Esse método será chamado com uma série de keyword arguments. Por enquando o único kwarg passado é
o `logger`. Nesse kwarg é passado uma instância de Logger para que o plugin possa usar.

Essa função deve retornar um dict com essa estrutura:

```python
{
  "blueprint": <blueprint-instance>
}
```

Onde `<blueprint-instance>` é uma instância de um `Flask.Blueprint`. Cada rota registrada nesse blueprint
será plugada na Asgard API abaixo do path: `/_cat/metrics/<package-name>/*`.

Aqui, `<package-name>` é o nome do seu pacote registrado no Pypi.

## Registrando sua função de inicialização

Em seu `setup.py` adicione esse trecho:

```python

    entry_points={
        'asgard_api_metrics_mountpoint': [
            'init = <module>:function_name',
        ],
    }
```

Dessa forma, quando a Asgard API der boot, ela vai chmar a função definida nessa configuração. Os logs as Asgard API indicam se
o carregamento dos plugin foi feito com sucesso ou não.

## Instalando seu plugin em sua Asgard API

Para que esse plugin seja ativado, basta que você faça `pip install asgard-api-plugin-metrics-mesos==<version>` durante o build da sua Asgard API, ou seja,
basta adicionar esse pacote (`asgard-api-plugin-metrics-mesos`) como dependencia no seu deploy da Asgard API.

# Endpoints implementados por esse plugin


## Routes:
* /attrs: Retorna uma lista com todos os atributos usados em todos os slaves do cluster. Atributos de nome igual são agrupados;
* /attrs/count: Retorna a contagem de quantos atributos estão sendo usados em todo o cluster;
* /slaves-with-attrs?**attr**=**value**: Retorna todos os slaves do cluster que possuem o atributo `attr` com valor `value`;
* /slaves-with-attrs/count?**attr**=**value**: Retorna a contagem de slaves que possuem o atributo `attr` com valor `value`;
* /attr-usage?**attr**=**value**: Retorna o uso (percentual e absoluto) de CPU e RAM dos slaves que possuem o atributo `attr` com o valor `value`;
* /master/<ip>?prefix=<prefix>: Retorna as métricas do Mesos master cpm IP=`<ip>` que começam por <prefix>;
* /slave/<ip>?prefix=<prefix>: Retorna as métricas do Mesos slave com IP=`<ip>` que começam por <prefix>;
* /leader?prefix=<prefix>: Igual aos endpoints acima, mas descobre quem é o atual lider e pega os dados dele;
* /masters/alive: Retorna um JSON com cada IP de master e uma indicação se eles responderam ao "ping", ex:

```
{
   "http://172.18.0.11:5050" : 0,
   "http://172.18.0.13:5050" : 1,
   "http://172.18.0.12:5050" : 1
}
```

## Env vars
* Todas as env são lidas pela `asgard-api-sdk`. As que são necessária aqui são 
as que possuem sufixo `_MESOS_ADDRESS_<N>`. Mais detalhes na doc da [asgard-api-sdk](https://github.com/B2W-BIT/asgard-api-sdk).


## Running tests:
`$ PYTHONPATH=. py.test --cov=./ --cov-report term-missing -v -s`


CHANGELOG.md
# Mesos Metrics (docker.sieve.com.br/mesos-metrics)

## Changelog


### 0.9.0
 * Adicionando chave `masters_down` no endpoinrt `masters/alive` para mostrar quantos masters estão sem responder.

### 0.6.0

 * Implementação do entpoint /tasks/count para podermos ter a contagem de tasks no cluster (IN-2535)

### 0.5.0
  - Implementação de endpoints `/attrs/count` e `/slaves-with-attrs/count`;
  - Implementação da leitura de métricas sempre do mesos leader;
  - Uso do asgard-api-sdk para ter acesso a envvars multi-valor.

### 0.4.1
  - Adição de logs de debug na chamada à API do mesos

### 0.4.0
  - Recebe logger que é passado pela API no momento de inicializar o plugin

### 0.3.0rc2
  - Novo endpoint para pegar métricas sempre do atual mesos master lider

### 0.3.0rc1
  - Novos endpoints para buscar métricas dos mesos master e slaves

### 0.2.0
  - Versão 0.1.0 estava quebrada. Faltou o o registro do plugin

### 0.1.0
  - Migração do projeto para ser um plugin do asgard-api



tests/__init__.py
import json
import os
from typing import Dict

CURRENT_DIR = os.path.dirname(__file__)

def get_fixture(file_name: str) -> Dict:
    with open(os.path.join(CURRENT_DIR, file_name)) as fp:
        return json.load(fp)

def with_json_fixture(fixture_path):
    def wrapper(func):
        def decorator(self):
            fixture = get_fixture(fixture_path)
            return func(self, fixture)
        return decorator
    return wrapper

SLAVES_STATE = {
    "slaves":[
        {
            "resources":{"disk":600000.0,"mem": 16000.0,"gpus":0.0,"cpus":4.0},
            "used_resources":{"disk":0.0,"mem":8000.0,"gpus":0.0,"cpus":2.4},
            "attributes":{"dc":"bit","server":"slave0"},
            "active": True
        },
        {
            "resources":{"disk":400000.0,"mem": 8000.0,"gpus":0.0,"cpus":8.0},
            "used_resources":{"disk":0.0,"mem":8000.0,"gpus":0.0,"cpus":3.6},
            "attributes":{"dc":"bit","server":"slave1"},
            "active": True
        },
    ]
}


tests/test_mesos.py
from unittest import TestCase, mock
import os
import importlib

from flask import Flask
from responses import RequestsMock
import json
import requests

from metrics import mesos, mesos_metrics_blueprint, config
import tests
from tests import with_json_fixture
import metrics

class MesosTest(TestCase):

    def setUp(self):
        self.application = Flask(__name__)
        self.application.register_blueprint(mesos_metrics_blueprint, url_prefix="/metrics")
        self.get_mesos_leader_address_patcher = mock.patch('asgard.sdk.mesos.get_mesos_leader_address', return_value="http://10.0.0.1:5050")
        self.get_mesos_leader_addres_mock = self.get_mesos_leader_address_patcher.start()

    def tearDown(self):
        self.get_mesos_leader_address_patcher.stop()

    def test_get_slaves_attr(self):
        attrs = mesos.get_slaves_attr(tests.SLAVES_STATE)
        self.assertEqual(
            set(attrs['server']),
            set(['slave0', 'slave1'])
        )
        self.assertEqual(
            set(attrs['dc']),
            set(['bit'])
        )

    def test_get_slaves_with_attr(self):
        result = mesos.get_slaves_with_attr(tests.SLAVES_STATE, {"dc": "bit"})
        self.assertEqual(len(result), 2)

    def test_get_slaves_with_attr_empty_result(self):
        result = mesos.get_slaves_with_attr(tests.SLAVES_STATE, {"foo": "bar"})
        self.assertEqual(result, [])

    def test_get_slaves_with_attr_filtered(self):
        result = mesos.get_slaves_with_attr(tests.SLAVES_STATE, {"server": "slave0"})
        self.assertEqual(result[0]['attributes']['server'], 'slave0')

    def test_attr_usage(self):
        with mock.patch('metrics.mesos.get_slaves_with_attr') as mocked:
            mocked.return_value = tests.SLAVES_STATE['slaves']
            result = mesos.get_attr_usage(tests.SLAVES_STATE, {"dc": "bit"})

            self.assertDictEqual(
                result,
                {
                    'cpu_used': 6.0,
                    'ram_used': 16,
                    'ram_total': 24,
                    'cpu_total': 12.0,
                    'ram_pct': 66.7,
                    'cpu_pct': 50.0
                }
            )

    @with_json_fixture("fixtures/master-slave-data.json")
    def test_attr_usage_no_usage(self, mesos_slaves_data):
        """
        Certificamos que conseguimos retonar um valor zero, ou seja,
        caso não exista nenhum slave usando a tag escolhida no filtro.
        """
        client = self.application.test_client()
        with RequestsMock() as rsps:

            rsps.add(method='GET', url="http://10.0.0.1:5050/slaves", body=json.dumps(mesos_slaves_data), status=200)
            response = client.get("/metrics/attr-usage?no-usage=none")
            self.assertEqual(200, response.status_code)

            self.assertDictEqual(
                json.loads(response.data),
                {
                    'cpu_used': 0.0,
                    'ram_used': 0,
                    'ram_total': 0,
                    'cpu_total': 0.0,
                    'ram_pct': 0.0,
                    'cpu_pct': 0.0
                }
            )

    def test_get_masters_alive_all_ok(self):
        meso_master_addresses = ["http://10.0.0.1:5050", "http://10.0.0.2:5050", "http://10.0.0.3:5050"]
        client = self.application.test_client()
        with RequestsMock() as rsps, \
                mock.patch('asgard.sdk.options', get_option=mock.MagicMock(return_value=meso_master_addresses)):
            importlib.reload(metrics)
            rsps.add(method='GET', url="http://10.0.0.1:5050/health", body="", status=200)
            rsps.add(method='GET', url="http://10.0.0.2:5050/health", body="", status=200)
            rsps.add(method='GET', url="http://10.0.0.3:5050/health", body="", status=200)
            response = client.get("/metrics/masters/alive")
            self.assertEqual(200, response.status_code)
            response_data = json.loads(response.data)
            self.assertEqual(1, response_data['http://10.0.0.1:5050'])
            self.assertEqual(1, response_data['http://10.0.0.2:5050'])
            self.assertEqual(1, response_data['http://10.0.0.3:5050'])
            self.assertEqual(1, response_data['all_ok'])
            self.assertEqual(0, response_data['masters_down'])

    def test_get_masters_alive_one_down(self):
        meso_master_addresses = ["http://10.0.0.1:5050", "http://10.0.0.2:5050", "http://10.0.0.3:5050"]
        client = self.application.test_client()
        with RequestsMock() as rsps, \
                mock.patch('asgard.sdk.options', get_option=mock.MagicMock(return_value=meso_master_addresses)):
            importlib.reload(metrics)
            rsps.add(method='GET', url="http://10.0.0.1:5050/health", body="", status=200)
            rsps.add(method='GET', url="http://10.0.0.2:5050/health", body=requests.exceptions.ConnectionError())
            rsps.add(method='GET', url="http://10.0.0.3:5050/health", body="", status=200)
            response = client.get("/metrics/masters/alive")
            self.assertEqual(200, response.status_code)
            response_data = json.loads(response.data)
            self.assertEqual(1, response_data['http://10.0.0.1:5050'])
            self.assertEqual(0, response_data['http://10.0.0.2:5050'])
            self.assertEqual(1, response_data['http://10.0.0.3:5050'])
            self.assertEqual(0, response_data['all_ok'])
            self.assertEqual(1, response_data['masters_down'])

    def test_get_masters_alive_all_down(self):
        meso_master_addresses = ["http://10.0.0.1:5050", "http://10.0.0.2:5050", "http://10.0.0.3:5050"]
        client = self.application.test_client()
        with RequestsMock() as rsps, \
                mock.patch('asgard.sdk.options', get_option=mock.MagicMock(return_value=meso_master_addresses)):
            importlib.reload(metrics)
            rsps.add(method='GET', url="http://10.0.0.1:5050/health", body=requests.exceptions.ConnectionError())
            rsps.add(method='GET', url="http://10.0.0.2:5050/health", body=requests.exceptions.ConnectionError())
            rsps.add(method='GET', url="http://10.0.0.3:5050/health", body=requests.exceptions.ConnectionError())
            response = client.get("/metrics/masters/alive")
            self.assertEqual(200, response.status_code)
            response_data = json.loads(response.data)
            self.assertEqual(0, response_data['http://10.0.0.1:5050'])
            self.assertEqual(0, response_data['http://10.0.0.2:5050'])
            self.assertEqual(0, response_data['http://10.0.0.3:5050'])
            self.assertEqual(0, response_data['all_ok'])
            self.assertEqual(3, response_data['masters_down'])

    @with_json_fixture("fixtures/master-slave-data.json")
    def test_get_slaves_with_attrs_count_endpoint(self, master_state_fixture):
        client = self.application.test_client()
        with RequestsMock() as rsps:
            rsps.add(method='GET', url="http://10.0.0.1:5050/slaves", body=json.dumps(master_state_fixture), status=200)
            response = client.get("/metrics/slaves-with-attrs/count?dc=bit")
            self.assertEqual(200, response.status_code)
            response_data = json.loads(response.data)
            self.assertEqual(6, response_data['total_slaves'])

    @with_json_fixture("fixtures/master-tasks.json")
    def test_get_tasks_count_endpoint(self, master_tasks_fixture):
        client = self.application.test_client()
        with RequestsMock() as rsps:
            rsps.add(method='GET', url="http://10.0.0.1:5050/tasks?limit=-1", body=json.dumps(master_tasks_fixture), status=200, match_querystring=True)
            response = client.get("/metrics/tasks/count")
            self.assertEqual(200, response.status_code)
            response_data = json.loads(response.data)
            self.assertEqual(2, response_data['task_running'])
            self.assertEqual(1, response_data['task_failed'])
            self.assertEqual(2, response_data['task_finished'])
            self.assertEqual(1, response_data['task_killed'])
            self.assertEqual(6, response_data['total'])


    def test_get_task_namespace_prefix(self):
        self.assertEqual("sieve_", mesos._get_task_namespace("sieve_app_other_name.UUID"))
        self.assertEqual("infra_", mesos._get_task_namespace("infra_app_other_name.UUID"))
        self.assertEqual("other_", mesos._get_task_namespace("ct:1528217280003:0:afiliados-sleep-10:"))


    @with_json_fixture("fixtures/master-tasks-multiple-namespaces.json")
    def test_get_tasks_count_endpoint_separate_namespaces(self, master_tasks_fixture):
        """
        Confirmamos que fazemos a contagem por namespace também
        """
        client = self.application.test_client()
        with RequestsMock() as rsps:
            rsps.add(method='GET', url="http://10.0.0.1:5050/tasks?limit=-1", body=json.dumps(master_tasks_fixture), status=200, match_querystring=True)
            response = client.get("/metrics/tasks/count")
            self.assertEqual(200, response.status_code)
            response_data = json.loads(response.data)
            self.assertEqual(2, response_data['task_running'])
            self.assertEqual(3, response_data['task_failed'])
            self.assertEqual(2, response_data['task_finished'])
            self.assertEqual(2, response_data['task_killed'])
            self.assertEqual(9, response_data['total'])

            # Namespace sieve
            self.assertEqual(1, response_data['sieve_task_running'])

            # Namespace infra
            self.assertEqual(0, response_data['infra_task_running'])

            # Chronos (que por enquanto não pertence a nehum namespace)
            self.assertEqual(1, response_data['other_task_running'])

    @with_json_fixture("fixtures/master-slave-data.json")
    def test_attrs_endpoint(self, master_state_fixture):
        client = self.application.test_client()
        with RequestsMock() as rsps:
            rsps.add(method='GET', url="http://10.0.0.1:5050/slaves", body=json.dumps(master_state_fixture), status=200)
            response = client.get("/metrics/attrs")
            self.assertEqual(200, response.status_code)
            response_data = json.loads(response.data)
            self.assertEqual(4, len(response_data))
            self.assertEqual(["dc", "mesos", "owner", "workload"], sorted(response_data.keys()))

    @with_json_fixture("fixtures/master-slave-data.json")
    def test_attrs_endpoint_count(self, master_state_fixture):
        client = self.application.test_client()
        with RequestsMock() as rsps:
            rsps.add(method='GET', url="http://10.0.0.1:5050/slaves", body=json.dumps(master_state_fixture), status=200)
            response = client.get("/metrics/attrs/count")
            self.assertEqual(200, response.status_code)
            response_data = json.loads(response.data)
            self.assertEqual(1, len(response_data))
            self.assertEqual(4, response_data['total_attrs'])

    @with_json_fixture("fixtures/master-metrics.json")
    def test_get_metrics_from_current_leader(self, leader_metrics_fixture):
        with self.application.test_client() as client:
            with RequestsMock() as rsps:
                rsps.add(method='GET', url="http://10.0.0.1:5050/metrics/snapshot", body=json.dumps(leader_metrics_fixture), status=200)
                response = client.get("/metrics/leader?prefix=sys")
                self.get_mesos_leader_addres_mock.assert_called()
                self.assertEqual(200, response.status_code)
                metrics_data = json.loads(response.data)
                self.assertEqual(6, len(metrics_data.keys()))
                self.assertEqual(8, metrics_data["system/cpus_total"])
                self.assertEqual(0.1, metrics_data["system/load_15min"])
                self.assertEqual(0.21, metrics_data["system/load_1min"])
                self.assertEqual(0.14, metrics_data["system/load_5min"])
                self.assertEqual(269631488, metrics_data["system/mem_free_bytes"])
                self.assertEqual(20935741440, metrics_data["system/mem_total_bytes"])

    @with_json_fixture("fixtures/master-metrics.json")
    def test_get_metrics_master_with_prefix(self, master_metrics_fixture):
        with self.application.test_client() as client:
            with RequestsMock() as rsps:
                rsps.add(method='GET', url="http://10.0.0.1:5050/metrics/snapshot", body=json.dumps(master_metrics_fixture), status=200)
                response = client.get("/metrics/master/10.0.0.1?prefix=sys")
                self.assertEqual(200, response.status_code)
                metrics_data = json.loads(response.data)
                self.assertEqual(6, len(metrics_data.keys()))
                self.assertEqual(8, metrics_data["system/cpus_total"])
                self.assertEqual(0.1, metrics_data["system/load_15min"])
                self.assertEqual(0.21, metrics_data["system/load_1min"])
                self.assertEqual(0.14, metrics_data["system/load_5min"])
                self.assertEqual(269631488, metrics_data["system/mem_free_bytes"])
                self.assertEqual(20935741440, metrics_data["system/mem_total_bytes"])

    @with_json_fixture("fixtures/master-metrics.json")
    def test_get_metrics_master_no_prefix(self, master_metrics_fixture):
        with self.application.test_client() as client:
            with RequestsMock() as rsps:
                rsps.add(method='GET', url="http://10.0.0.1:5050/metrics/snapshot", body=json.dumps(master_metrics_fixture), status=200)
                response = client.get("/metrics/master/10.0.0.1")
                self.assertEqual(200, response.status_code)
                metrics_data = json.loads(response.data)
                self.assertEqual(len(master_metrics_fixture.keys()), len(metrics_data.keys()))

    @with_json_fixture("fixtures/slave-metrics.json")
    def test_get_metrics_slave_no_prefix(self, slave_metrics_fixture):
        with self.application.test_client() as client:
            with RequestsMock() as rsps:
                rsps.add(method='GET', url="http://10.0.0.1:5051/metrics/snapshot", body=json.dumps(slave_metrics_fixture), status=200)
                response = client.get("/metrics/slave/10.0.0.1")
                self.assertEqual(200, response.status_code)
                metrics_data = json.loads(response.data)
                self.assertEqual(len(slave_metrics_fixture.keys()), len(metrics_data.keys()))

    @with_json_fixture("fixtures/slave-metrics.json")
    def test_get_metrics_slave_with_prefix(self, slave_metrics_fixture):
        with self.application.test_client() as client:
            with RequestsMock() as rsps:
                rsps.add(method='GET', url="http://10.0.0.1:5050/metrics/snapshot", body=json.dumps(slave_metrics_fixture), status=200)
                response = client.get("/metrics/leader?prefix=containerizer/mesos/provisioner/")
                self.assertEqual(200, response.status_code)
                metrics_data = json.loads(response.data)
                self.assertEqual(2, len(metrics_data.keys()))
                self.assertEqual(0, metrics_data["containerizer/mesos/provisioner/bind/remove_rootfs_errors"])
                self.assertEqual(0, metrics_data["containerizer/mesos/provisioner/remove_container_errors"])



tests/test_plugin_init.py
import unittest
import logging

from metrics import asgard_api_plugin_init, config


class TestPluginInit(unittest.TestCase):

    def test_read_logger_if_passed(self):
        logger = logging.getLogger("logger.test")
        asgard_api_plugin_init(logger=logger)
        self.assertTrue(logger is config.logger)

    def test_default_logger_is_not_none(self):
        asgard_api_plugin_init()
        self.assertTrue(config.logger is not None)


entrypoint.py
from flask import Blueprint, Response, request
from metrics import mesos, config
from json import dumps

app = Blueprint(__name__, __name__)

@app.route('/attrs')
def attrs():
    slaves_state = mesos.get_mesos_slaves()
    return Response(
        dumps(mesos.get_slaves_attr(slaves_state)),
        mimetype='application/json'
    )

@app.route('/slaves-with-attrs')
def slaves_with_attrs():
    slaves_state = mesos.get_mesos_slaves()
    result = mesos.get_slaves_with_attr(
        slaves_state,
        dict(request.args.items())
    )
    return Response(
        dumps(result),
        mimetype='application/json'
    )

@app.route('/attr-usage')
def slaves_attr_usage():
    slaves_state = mesos.get_mesos_slaves()
    result = mesos.get_attr_usage(
        slaves_state,
        dict(request.args.items())
    )
    return Response(
        dumps(result),
        mimetype='application/json'
    )



metrics/util.py

import requests

from asgard.sdk import mesos

from metrics import config

def get_mesos_slaves():
    leader_address = mesos.get_mesos_leader_address()
    url = f"{leader_address}/slaves"
    config.logger.debug({"action": "pre-fetch", "fetch-url": url})
    response = requests.get(url, timeout=2)
    config.logger.debug({"action": "post-fetch", "fetch-url": url, "fetch-status": response.status_code})
    return response.json()

def get_mesos_tasks():
    leader_address = mesos.get_mesos_leader_address()
    url = f"{leader_address}/tasks?limit=-1"
    config.logger.debug({"action": "pre-fetch", "fetch-url": url})
    response = requests.get(url, timeout=2)
    config.logger.debug({"action": "post-fetch", "fetch-url": url, "fetch-status": response.status_code})
    return response.json()


metrics/mesos.py
from metrics import config
from metrics.util import get_mesos_slaves, get_mesos_tasks
from requests import get
from collections import defaultdict

CHRONOS_TASK_PREFIX = "ct"
MESOS_TASK_RUNNING = "TASK_RUNNING"

def _get_task_namespace(task_name):
    ns = task_name.split("_")[0]
    if task_name.startswith(CHRONOS_TASK_PREFIX):
        ns = "other"
    return f"{ns}_"

def get_tasks():
    all_tasks = get_mesos_tasks()
    counts = defaultdict(int)
    for task in all_tasks['tasks']:
        ns = _get_task_namespace(task['id'])
        counts['total'] += 1
        counts[task['state'].lower()] += 1

        state = task['state']
        if state == MESOS_TASK_RUNNING:
            counts[f"{ns}{state.lower()}"] += 1

        counts[f"{ns}{MESOS_TASK_RUNNING.lower()}"]

    return counts

def get_slaves_attr(slaves_state):
    attrs = defaultdict(set)

    for slave in slaves_state['slaves']:
        for attr in slave['attributes']:
            attrs[attr].add(slave['attributes'][attr])

    for attr in attrs:
        attrs[attr] = list(attrs[attr])

    return dict(attrs)

def get_slaves_with_attr(slaves_state, attrs):
    slaves = []

    for slave in slaves_state['slaves']:
        intersection = dict(set(slave['attributes'].items()).intersection(set(attrs.items())))

        if intersection != attrs:
            continue

        slaves.append(slave)

    return slaves

def get_attr_usage(slaves_state, attrs):
    slaves = get_slaves_with_attr(slaves_state, attrs)

    usage = 0

    cpu_total = 0
    ram_total = 0

    cpu_used = 0
    ram_used = 0

    for slave in slaves:
        cpu_total = cpu_total + slave["resources"]['cpus']
        ram_total = ram_total + slave["resources"]['mem']

        cpu_used = cpu_used + slave["used_resources"]['cpus']
        ram_used = ram_used + slave["used_resources"]['mem']

    ram_total = round(ram_total/1000)
    ram_used = round(ram_used/1000)

    return {
        'cpu_total': cpu_total,
        'ram_total': ram_total,
        'cpu_used': cpu_used,
        'ram_used': ram_used,
        'cpu_pct': round(cpu_used*100/(cpu_total or 1), 1),
        'ram_pct': round(ram_used*100/(ram_total or 1), 1)
    }


metrics/__init__.py
from json import dumps
import json

import requests
from flask import Blueprint, Response, request

from metrics import mesos, config
from asgard.sdk import mesos as mesos_asgard_sdk
from asgard.sdk import options

mesos_metrics_blueprint = Blueprint(__name__, __name__)

def asgard_api_plugin_init(**kwargs):
    config.logger = kwargs.get("logger", config.logger)
    return {
        'blueprint': mesos_metrics_blueprint
    }

@mesos_metrics_blueprint.route('/attrs/count')
def attrs_count():
    config.logger.debug("Counting attrs")
    slaves_state = mesos.get_mesos_slaves()
    slaves_attrs = mesos.get_slaves_attr(slaves_state)
    return Response(
        dumps({'total_attrs': len(slaves_attrs.keys())}),
        mimetype='application/json'
    )

@mesos_metrics_blueprint.route('/tasks/count')
def tasks_count():
    config.logger.debug("Counting tasks")
    mesos_tasks = mesos.get_tasks()
    return Response(
        dumps(mesos_tasks),
        mimetype='application/json'
    )

@mesos_metrics_blueprint.route('/attrs')
def attrs():
    config.logger.debug("Reading attrs")
    slaves_state = mesos.get_mesos_slaves()
    return Response(
        dumps(mesos.get_slaves_attr(slaves_state)),
        mimetype='application/json'
    )


@mesos_metrics_blueprint.route('/slaves-with-attrs/count')
def slaves_with_attrs_count():
    slaves_state = mesos.get_mesos_slaves()
    result = mesos.get_slaves_with_attr(
        slaves_state,
        dict(request.args.items())
    )
    return Response(
        dumps({'total_slaves': len(result)}),
        mimetype='application/json'
    )

@mesos_metrics_blueprint.route('/slaves-with-attrs')
def slaves_with_attrs():
    slaves_state = mesos.get_mesos_slaves()
    result = mesos.get_slaves_with_attr(
        slaves_state,
        dict(request.args.items())
    )
    return Response(
        dumps(result),
        mimetype='application/json'
    )

@mesos_metrics_blueprint.route('/attr-usage')
def slaves_attr_usage():
    slaves_state = mesos.get_mesos_slaves()
    result = mesos.get_attr_usage(
        slaves_state,
        dict(request.args.items())
    )
    return Response(
        dumps(result),
        mimetype='application/json'
    )

@mesos_metrics_blueprint.route("/masters/alive")
def masters_alive():
    mesos_addresses = options.get_option("MESOS", "ADDRESS")
    result = {addr: int(mesos_asgard_sdk.is_master_healthy(addr)) for addr in mesos_addresses}
    all_ok = all([result[key] for key in result.keys()])
    masters_down = len(list(filter(lambda r: r[1] == 0, result.items())))

    result['all_ok'] = int(all_ok)
    result['masters_down'] = masters_down
    return Response(
        dumps(result),
        mimetype='application/json'
    )



def filter_mesos_metrics(server_address, prefix):
    metrics_data = requests.get(f"{server_address}/metrics/snapshot").json()
    filtered_metrics = {}
    for data_key, data_value in metrics_data.items():
        if data_key.startswith(prefix):
            filtered_metrics[data_key] = data_value
    return filtered_metrics

@mesos_metrics_blueprint.route("/leader")
def leader_metrics():
    prefix = request.args.get("prefix", "")
    server_address = mesos_asgard_sdk.get_mesos_leader_address()
    filtered_metrics = filter_mesos_metrics(server_address, prefix)
    return Response(json.dumps(filtered_metrics), mimetype='application/json')

@mesos_metrics_blueprint.route("/master/<string:server_ip>")
def master_metrics(server_ip):
    prefix = request.args.get("prefix", "")
    filtered_metrics = filter_mesos_metrics(f"http://{server_ip}:5050", prefix)
    return Response(json.dumps(filtered_metrics), mimetype='application/json')

@mesos_metrics_blueprint.route("/slave/<string:server_ip>")
def slave_metrics(server_ip):
    prefix = request.args.get("prefix", "")
    filtered_metrics = filter_mesos_metrics(f"http://{server_ip}:5051", prefix)
    return Response(json.dumps(filtered_metrics), mimetype='application/json')

